File: python/day16.py
#!/usr/bin/env python3

fields = []
my_ticket = None
nearby_tickets = []

with open("../input/day16.txt") as f:
    # departure location: 43-237 or 251-961
    while True:
        l = f.readline().strip()
        if not l:
            break
        field, rstring = l.split(': ')
        fields.append([field, []])
        rstrings = rstring.split(' or ')
        for rstring in rstrings:
            min_max = list(map(int, rstring.split('-')))
            fields[-1][1].append(min_max)
    # your ticket:
    f.readline()
    # 151,71,67,113,127,163,131,59,137,103,73,139,107,101,97,149,157,53,109,61
    while True:
        l = f.readline().strip()
        if not l:
            break
        my_ticket = list(map(int, l.split(',')))
    # nearby tickets:
    f.readline()
    # 680,418,202,55,792,800,896,801,312,252,721,702,24,112,608,837,98,222,797,364
    while True:
        l = f.readline().strip()
        if not l:
            break
        nearby_tickets.append(list(map(int, l.split(','))))

error_rate = 0
for ticket in nearby_tickets:
    for val in ticket:
        val_match = False
        for name, minmaxes in fields:
            for minmax in minmaxes:
                if minmax[0] <= val <= minmax[1]:
                    val_match = True
                    break
            if val_match:
                break
        if not val_match:
            error_rate += val
print(error_rate)

File: c++/day16.cpp
#include <fstream>
#include <iostream>
#include <map>
#include <string>
#include <vector>

struct FieldRange {
    int min;
    int max;
};
struct FieldDef {
    std::string name;
    std::vector<FieldRange> ranges;
};
using Ticket = std::vector<int>;
struct Input {
    std::vector<FieldDef> fieldDefs;
    Ticket yourTicket;
    std::vector<Ticket> nearbyTickets;
};

// 680,418,202,55,792,800,896,801,312,252,721,702,24,112,608,837,98,222,797,364
void parseTicket(std::string line, Ticket &ticket) {
    std::string::size_type parsePos = 0;
    while (parsePos != std::string::npos) {
        ticket.emplace_back(atoi(&line[parsePos]));
        parsePos = line.find(',', parsePos);
        if (parsePos != std::string::npos)
            parsePos++;
    }
}

Input readParseInput(std::string fileName) {
    std::ifstream file(fileName);
    Input input;
    std::string line;

    // list of: departure location: 43-237 or 251-961
    while (std::getline(file, line)) {
        if (!line.length())
            break;
        FieldDef *fd = &input.fieldDefs.emplace_back();

        std::string::size_type parsePos = 0;
        std::string::size_type colonPos = line.find(':', parsePos);
        fd->name = line.substr(parsePos, colonPos - parsePos);
        parsePos = colonPos + 2;

        while (parsePos != std::string::npos) {
            FieldRange *fr = &fd->ranges.emplace_back();

            fr->min = atoi(&line[parsePos]);
            parsePos = line.find('-', parsePos) + 1;

            fr->max = atoi(&line[parsePos]);
            parsePos = line.find(' ', parsePos);
            if (parsePos != std::string::npos)
                parsePos += 4; // ' or '
        }
    }

    // your ticket:
    std::getline(file, line);
    // ticket
    std::getline(file, line);
    parseTicket(line, input.yourTicket);

    // blank line
    std::getline(file, line);

    // nearby tickets:
    std::getline(file, line);
    // list of: ticket
    while (std::getline(file, line)) {
        parseTicket(line, input.nearbyTickets.emplace_back());
    }

    file.close();

    return input;
}

int answer(const Input &input) {
    int result = 0;

    for (const Ticket &t : input.nearbyTickets) {
        for (int fieldVal : t) {
            bool validForSomeFieldDef = false;
            for (const FieldDef &fd : input.fieldDefs) {
                for (const FieldRange &fr : fd.ranges) {
                    if ((fr.min <= fieldVal) && (fieldVal <= fr.max)) {
                        validForSomeFieldDef = true;
                        break;
                    }
                }
                if (validForSomeFieldDef)
                    break;
            }
            if (!validForSomeFieldDef)
                result += fieldVal;
        }
    }

    return result;
}

int main(void) {
    std::cout << answer(readParseInput("../input/day16.txt")) << '\n';
    return 0;
}
